File: frontback/src/main/kotlin/com/digitalsamurai/opentelemetry/example/EnvironmentData.kt
package com.digitalsamurai.opentelemetry.example

data class EnvironmentData(
    val server: ServerData,
    val services: Services,
) {
    data class ServerData(
        val host: String,
        val port: Int
    )

    data class Services(
        val otel: Otel,
        val auth: Auth
    ) {
        data class Otel(
            val host: String,
            val port: Int,
        )

        data class Auth(
            val host: String,
            val port: Int,
        )
    }

    companion object {
        fun fromArgsOrSystemEnv(args: Array<String>): EnvironmentData {
            val serverHost = args.getOrNull(args.indexOf("server_host") + 1) ?: System.getenv("server_host")
            val serverPort = (args.getOrNull(args.indexOf("server_port") + 1) ?: System.getenv("server_port")).toInt()

            val otelHost = args.getOrNull(args.indexOf("otel_host") + 1) ?: System.getenv("otel_host")
            val otelPort = (args.getOrNull(args.indexOf("otel_port") + 1) ?: System.getenv("otel_port")).toInt()

            val authHost = args.getOrNull(args.indexOf("auth_host") + 1) ?: System.getenv("auth_host")
            val authPort = (args.getOrNull(args.indexOf("auth_port") + 1) ?: System.getenv("auth_port")).toInt()

            return EnvironmentData(
                server = ServerData(
                    host = serverHost,
                    port = serverPort
                ),
                services = Services(
                    otel = Services.Otel(
                        host = otelHost,
                        port = otelPort
                    ),
                    auth = Services.Auth(
                        host = authHost,
                        port = authPort
                    )
                ),
            )
        }
    }
}

File: backend/environment/environment_data.go
package environment

import (
	"flag"
	"log"
	"os"
	"strconv"
)

type Data struct {
	Host     string
	Port     int
	Services Services
}

type Services struct {
	Auth AuthService
	Otel OtelService
}

type AuthService struct {
	Key    string
	Issuer string
}

type OtelService struct {
	Host string
	Port int
}

func LoadEnvironmentData() *Data {
	serverPort, err := strconv.Atoi(os.Getenv("SERVER_PORT"))
	if err != nil {
		// TODO локально юзаются аргументы, поэтому нельзя кидать фатал
	}
	otelPort, err := strconv.Atoi(os.Getenv("OTEL_PORT"))
	if err != nil {
		// TODO локально юзаются аргументы, поэтому нельзя кидать фатал
	}
	data := &Data{
		Host: os.Getenv("SERVER_HOST"),
		Port: serverPort,
		Services: Services{
			Auth: AuthService{
				Key:    os.Getenv("AUTH_KEY"),
				Issuer: os.Getenv("AUTH_ISSUER"),
			},
			Otel: OtelService{
				Host: os.Getenv("OTEL_HOST"),
				Port: otelPort,
			},
		},
	}
	flag.StringVar(&data.Host, "server_host", data.Host, "Server host address")
	flag.IntVar(&data.Port, "server_port", data.Port, "Server serverPort address")

	flag.StringVar(&data.Services.Auth.Key, "auth_key", data.Services.Auth.Key, "Special secret key for auth")
	flag.StringVar(&data.Services.Auth.Issuer, "auth_issuer", data.Services.Auth.Issuer, "service")

	flag.StringVar(&data.Services.Otel.Host, "otel_host", data.Services.Otel.Host, "Otel host address")
	flag.IntVar(&data.Services.Otel.Port, "otel_port", data.Services.Otel.Port, "Otel grpc port")

	flag.Parse()
	if data.Host == "" || data.Port == 0 {
		log.Fatal("Environment data not parsed")
	}
	return data
}
